File: gui/src/discoverymanager.cpp
/*
 * This file is part of Chiaki.
 *
 * Chiaki is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chiaki is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chiaki.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <discoverymanager.h>
#include <exception.h>

#include <cstring>
#include <netinet/in.h>

#define PING_MS		500
#define HOSTS_MAX	16
#define DROP_PINGS	3

HostMAC DiscoveryHost::GetHostMAC() const
{
	QByteArray data = QByteArray::fromHex(host_id.toUtf8());
	if(data.size() != 6)
		return HostMAC();
	return HostMAC((uint8_t *)data.constData());
}

static void DiscoveryServiceHostsCallback(ChiakiDiscoveryHost *hosts, size_t hosts_count, void *user);

DiscoveryManager::DiscoveryManager(QObject *parent) : QObject(parent)
{
	chiaki_log_init(&log, CHIAKI_LOG_ALL & ~CHIAKI_LOG_VERBOSE, chiaki_log_cb_print, nullptr);

	service_active = false;
}

DiscoveryManager::~DiscoveryManager()
{
	if(service_active)
		chiaki_discovery_service_fini(&service);
}

void DiscoveryManager::SetActive(bool active)
{
	if(service_active == active)
		return;
	service_active = active;

	if(active)
	{
		ChiakiDiscoveryServiceOptions options;
		options.ping_ms = PING_MS;
		options.hosts_max = HOSTS_MAX;
		options.host_drop_pings = DROP_PINGS;
		options.cb = DiscoveryServiceHostsCallback;
		options.cb_user = this;

		sockaddr_in addr = {};
		addr.sin_family = AF_INET;
		addr.sin_port = htons(CHIAKI_DISCOVERY_PORT);
		addr.sin_addr.s_addr = 0xffffffff; // 255.255.255.255
		options.send_addr = reinterpret_cast<sockaddr *>(&addr);
		options.send_addr_size = sizeof(addr);

		ChiakiErrorCode err = chiaki_discovery_service_init(&service, &options, &log);
		if(err != CHIAKI_ERR_SUCCESS)
		{
			CHIAKI_LOGE(&log, "DiscoveryManager failed to init Discovery Service");
			return;
		}
	}
	else
	{
		chiaki_discovery_service_fini(&service);
		hosts = {};
		emit HostsUpdated();
	}

}

void DiscoveryManager::SendWakeup(const QString &host, const QByteArray &regist_key)
{
	addrinfo *addrinfos;
	int r = getaddrinfo(host.toLocal8Bit().constData(), nullptr, nullptr, &addrinfos);
	if(r != 0)
	{
		CHIAKI_LOGE(&log, "DiscoveryManager failed to getaddrinfo for wakeup");
		throw Exception("Failed to getaddrinfo");
	}
	sockaddr addr = {};
	socklen_t addr_len = 0;
	for(addrinfo *ai=addrinfos; ai; ai=ai->ai_next)
	{
		if(ai->ai_family != AF_INET)
			continue;
		if(ai->ai_protocol != IPPROTO_UDP)
			continue;
		if(ai->ai_addrlen > sizeof(addr))
			continue;
		std::memcpy(&addr, ai->ai_addr, ai->ai_addrlen);
		addr_len = ai->ai_addrlen;
		break;
	}
	freeaddrinfo(addrinfos);

	if(!addr_len)
	{
		CHIAKI_LOGE(&log, "DiscoveryManager failed to get suitable address from getaddrinfo for wakeup");
		throw Exception("Failed to get addr from getaddrinfo");
	}

	((sockaddr_in *)&addr)->sin_port = htons(CHIAKI_DISCOVERY_PORT);

	QByteArray key = regist_key;
	for(size_t i=0; i<key.size(); i++)
	{
		if(!key.at(i))
		{
			key.resize(i);
			break;
		}
	}

	bool ok;
	uint64_t credential = (uint64_t)QString::fromUtf8(key).toULongLong(&ok, 16);
	if(key.size() > 8 || !ok)
	{
		CHIAKI_LOGE(&log, "DiscoveryManager got invalid regist key for wakeup");
		throw Exception("Invalid regist key");
	}

	ChiakiDiscoveryPacket packet = {};
	packet.cmd = CHIAKI_DISCOVERY_CMD_WAKEUP;
	packet.user_credential = credential;

	ChiakiErrorCode err;
	if(service_active)
		err = chiaki_discovery_send(&service.discovery, &packet, &addr, addr_len);
	else
	{
		ChiakiDiscovery discovery;
		err = chiaki_discovery_init(&discovery, &log, AF_INET);
		if(err != CHIAKI_ERR_SUCCESS)
			throw Exception(QString("Failed to init Discovery: %1").arg(chiaki_error_string(err)));
		err = chiaki_discovery_send(&discovery, &packet, &addr, addr_len);
		chiaki_discovery_fini(&discovery);
	}

	if(err != CHIAKI_ERR_SUCCESS)
		throw Exception(QString("Failed to send Packet: %1").arg(chiaki_error_string(err)));
}

void DiscoveryManager::DiscoveryServiceHosts(QList<DiscoveryHost> hosts)
{
	this->hosts = std::move(hosts);
	emit HostsUpdated();
}

class DiscoveryManagerPrivate
{
	public:
		static void DiscoveryServiceHosts(DiscoveryManager *discovery_manager, const QList<DiscoveryHost> &hosts)
		{
			QMetaObject::invokeMethod(discovery_manager, "DiscoveryServiceHosts", Qt::ConnectionType::QueuedConnection, Q_ARG(QList<DiscoveryHost>, hosts));
		}
};

static void DiscoveryServiceHostsCallback(ChiakiDiscoveryHost *hosts, size_t hosts_count, void *user)
{
	QList<DiscoveryHost> hosts_list;
	hosts_list.reserve(hosts_count);

	for(size_t i=0; i<hosts_count; i++)
	{
		ChiakiDiscoveryHost *h = hosts + i;
		DiscoveryHost o = {};
		o.state = h->state;
		o.host_request_port = o.host_request_port;
#define CONVERT_STRING(name) if(h->name) { o.name = QString::fromLocal8Bit(h->name); }
		CHIAKI_DISCOVERY_HOST_STRING_FOREACH(CONVERT_STRING)
#undef CONVERT_STRING
		hosts_list.append(o);
	}

	DiscoveryManagerPrivate::DiscoveryServiceHosts(reinterpret_cast<DiscoveryManager *>(user), hosts_list);
}

File: android/app/src/main/java/com/metallic/chiaki/discovery/DiscoveryManager.kt
/*
 * This file is part of Chiaki.
 *
 * Chiaki is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chiaki is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chiaki.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.metallic.chiaki.discovery

import android.util.Log
import com.metallic.chiaki.common.MacAddress
import com.metallic.chiaki.common.ext.hexToByteArray
import com.metallic.chiaki.lib.CreateError
import com.metallic.chiaki.lib.DiscoveryHost
import com.metallic.chiaki.lib.DiscoveryService
import com.metallic.chiaki.lib.DiscoveryServiceOptions
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.Subject
import java.lang.NumberFormatException
import java.net.InetSocketAddress
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit

val DiscoveryHost.ps4Mac get() = this.hostId?.hexToByteArray()?.let {
	if(it.size == MacAddress.LENGTH)
		MacAddress(it)
	else
		null
}

class DiscoveryManager
{
	companion object
	{
		const val HOSTS_MAX: ULong = 16U
		const val DROP_PINGS: ULong = 3U
		const val PING_MS: ULong = 500U
		const val PORT = 987

		const val DEBOUNCE_EMPTY_MS = 1000L
	}

	private var discoveryService: DiscoveryService? = null

	private val discoveryActiveSubject: Subject<Boolean> = BehaviorSubject.create<Boolean>().also { it.onNext(false) }
	val discoveryActive: Observable<Boolean> get() = discoveryActiveSubject
	var active = false
		set(value)
		{
			field = value
			discoveryActiveSubject.onNext(value)
			updateService()
		}
	private var paused = false

	private val disposable = CompositeDisposable()

	private var discoveredHostsSubjectDebounced: Subject<List<DiscoveryHost>> = BehaviorSubject.create<List<DiscoveryHost>>().also {
		it.onNext(listOf())
	}.toSerialized()

	private var discoveredHostsSubjectRaw: Subject<List<DiscoveryHost>> = BehaviorSubject.create<List<DiscoveryHost>>().also { subject ->
		subject.debounce { hosts ->
				if(hosts.isEmpty())
					Observable.timer(DEBOUNCE_EMPTY_MS, TimeUnit.MILLISECONDS)
				else
					Observable.empty()
			}
			.subscribe { hosts ->
				discoveredHostsSubjectDebounced.onNext(hosts)
			}
			.addTo(disposable)
	}

	val discoveredHosts: Observable<List<DiscoveryHost>> get() = discoveredHostsSubjectDebounced

	fun resume()
	{
		paused = false
		updateService()
	}

	fun pause()
	{
		paused = true
		updateService()
	}

	fun dispose()
	{
		active = false
		disposable.dispose()
	}

	fun sendWakeup(host: String, registKey: ByteArray)
	{
		val registKeyString = registKey.indexOfFirst { it == 0.toByte() }.let { end -> registKey.copyOfRange(0, if(end >= 0) end else registKey.size) }.toString(StandardCharsets.UTF_8)
		val credential = try { registKeyString.toULong(16) } catch(e: NumberFormatException) {
			Log.e("DiscoveryManager", "Failed to convert registKey to int", e)
			return
		}
		DiscoveryService.wakeup(discoveryService, host, credential)
	}

	private fun updateService()
	{
		if(active && !paused && discoveryService == null)
		{
			discoveredHostsSubjectRaw.onNext(listOf())
			try
			{
				discoveryService = DiscoveryService(DiscoveryServiceOptions(
					HOSTS_MAX, DROP_PINGS, PING_MS, InetSocketAddress("255.255.255.255", PORT)
				), discoveredHostsSubjectRaw::onNext)
			}
			catch(e: CreateError)
			{
				Log.e("DiscoveryManager", "Failed to start Discovery Service: $e")
			}
		}
		else if((!active || paused) && discoveryService != null)
		{
			val service = discoveryService ?: return
			service.dispose()
			discoveryService = null
			if(!active)
				discoveredHostsSubjectRaw.onNext(listOf())
		}
	}
}